persist/encoding/decoder.go

```diff
@@ -42,3 +51,110 @@ type Decoder interface {
 	// DecodeLogEntry decodes commit log entry
 	DecodeLogEntry() (schema.LogEntry, error)
 }
+
+// DecoderStream is a data stream that is read by the decoder,
+// it takes both a reader and the underlying backing bytes.
+// This is constructed this way since the decoder needs access
+// to the backing bytes when taking refs directly for decoding byte
+// slices without allocating bytes itself but also needs to progress
+// the reader (for instance when a reader is a ReaderWithDigest that
+// is calculating a digest as its being read).
+type DecoderStream interface {
+	io.Reader
+
+	Reset(b []byte)
+
+	// Bytes returns the ref to the backing bytes, note decoders will
+	// call Skip if they "read" any bytes by simply taking refs to them
+	// so it is guaranteed that Read/Skip is called until EOF is reached.
+	Bytes() []byte
+
+	// Skip progresses the reader by a certain amount of bytes, useful
+	// when taking a ref to some of the bytes and progressing the reader
+	// itself.
+	Skip(length int64) error
+
+	// Remaining returns the remaining bytes in the stream.
+	Remaining() int64
+}
+
+type decoderStream struct {
+	reader       *bytes.Reader
+	bytes        []byte
+	lastReadByte int
```

Comment:
Why do we keep track of the last unread byte / read byte manually instead of using the APIs provided by `bytes.Reader`?

Reply:
`bytes.Reader` does not offer that API, only `bufio.Reader` does but we can't use it as we need to know the position of where into the buffer we are ourselves explicitly.